### PR TITLE
Colocalization module

### PR DESCRIPTION
**Pull request by tdrose**
Colocalization is a common metric for analysis of imaging MS/spatial metabolomics data.
In a previous publication from Theo ([ColocML](https://academic.oup.com/bioinformatics/article/36/10/3215/5734647)), they evaluated metrics/processing steps for computing colocalization. I implemented the best-performing metric.

I think it is a useful addition to the package for users to perform this kind of analysis/processing which cannot be done by other downstream analysis packages such as scanpy or squidpy.

### CODE REVIEW

**Comment by tdrose** [comment]
Once merged, closes #16 .

File: metaspace_converter/colocalization.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        layer: Key for ``adata.layer`` from which the ionimage_data for preprocessing taken.
+            If ``None``, ``adata.X`` is used. ``colocML_preprocessing`` will save the preprocessed
+            data per default in ``adata.layer['colocml_preprocessing']``.
+
+    Returns:
+        None. The processed data is saved in ``adata.uns['colocalization']``.
+    """
+
+    # Select data
+    if layer == None or layer not in adata.layers.keys():
```
[the remark this answers is not in the document]
Comment (tdrose) [reply metaspace_converter/colocalization.py:92]: Agree. Now raising an error if layer not available.

File: metaspace_converter/colocalization.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        data = np.array(adata.X).transpose()
+    else:
+        data = np.array(adata.layers[layer]).transpose()
+
+    # Compute colocalization
+    coloc = _pairwise_cosine_similarity(data)
+
+    # Save colocalization
+    adata.uns[COLOCALIZATION] = coloc
```
[the remark this answers is not in the document]
Comment (tdrose) [reply metaspace_converter/colocalization.py:101]: True. Good idea. Will move it to `anndata.varp`

File: docs/examples.rst [hunk 3]
```diff
@@ -114,3 +114,35 @@ Here using a reversed colormap which better represents intense values on bright
...
+
+   # Download data
+   adata = metaspace_to_anndata(dataset_id="2023-11-14_21h58m39s", fdr=0.1)
+
+   # Perform median filtering and quantile thresholding
+   # The processed data is saved as a layer `adata.layers["colocml_preprocessing"]`
+   # It has the same dimensions as `adata.X`
+   colocalization.colocML_preprocessing(adata, layer="colocml_preprocessing")
+
+   # Compute the pairwise colocalization metrix between all ion images
```
[the remark this answers is not in the document]
Comment (tdrose) [reply docs/examples.rst:141]: Done.

File: metaspace_converter/colocalization.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+import numpy as np
+from anndata import AnnData
+from scipy import ndimage
+
+from metaspace_converter.anndata_to_array import anndata_to_image_array
+from metaspace_converter.constants import COLOCALIZATION
+
+
+def colocML_preprocessing(
```
[the remark this answers is not in the document]
Comment (tdrose) [reply metaspace_converter/colocalization.py:11]: Done.

File: metaspace_converter/tests/colocalization_test.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            adata, median_filter_size=(3, 3), quantile_threshold=0.5, layer=COLOCML_LAYER
+        )
+
+        assert all(np.sum(adata.layers[COLOCML_LAYER] == 0, axis=0) <= 0.5 * adata.X.shape[0])
+
+        # Layer exists
+        assert COLOCML_LAYER in adata.layers.keys()
+
+        # Layer sizes match
+        assert adata.X.shape == adata.layers[COLOCML_LAYER].shape
```
[the remark this answers is not in the document]
Comment (tdrose) [reply metaspace_converter/tests/colocalization_test.py:90]: Done.

File: metaspace_converter/tests/colocalization_test.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        expected_preprocessing = np.median(actual[0][:3, :3])
+        observed = adata.layers[COLOCML_LAYER][adata.uns[METASPACE_KEY]["image_size"][X] + 1, 0]
+        assert observed == expected_preprocessing
+
+        expected_preprocessing = np.median(actual[1][:3, :3])
+        observed = adata.layers[COLOCML_LAYER][adata.uns[METASPACE_KEY]["image_size"][X] + 1, 1]
+        assert observed == expected_preprocessing
+
+        # Quantile thresholding
+        adata.X[0].reshape(-1)
```
[the remark this answers is not in the document]
Comment (tdrose) [reply metaspace_converter/tests/colocalization_test.py:84]: True. Probably just an artifact from my internal testing. Removed it.

File: metaspace_converter/tests/colocalization_test.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    )
+    return adata
+
+
+@pytest.mark.parametrize(
+    "adata_dummy",
+    [
+        dict(num_ions=5, height=10, width=10),
+        dict(num_ions=0, height=2, width=3),
+        dict(num_ions=4, height=4, width=4),
```
[the remark this answers is not in the document]
Comment (tdrose) [reply metaspace_converter/tests/colocalization_test.py:43]: Good idea. Done.

File: metaspace_converter/tests/colocalization_test.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        assert observed == expected_preprocessing
+
+        # Quantile thresholding
+        adata.X[0].reshape(-1)
+
+        colocML_preprocessing(
+            adata, median_filter_size=(3, 3), quantile_threshold=0.5, layer=COLOCML_LAYER
+        )
+
+        assert all(np.sum(adata.layers[COLOCML_LAYER] == 0, axis=0) <= 0.5 * adata.X.shape[0])
```
[the remark this answers is not in the document]
Comment (tdrose) [reply metaspace_converter/tests/colocalization_test.py]: Done. Added quantile_threshold to parameterization.

File: metaspace_converter/anndata_to_array.py [hunk 9]
```diff
@@ -16,13 +17,15 @@ def _check_pixel_coordinates(adata: AnnData) -> bool:
     return np.all(np.equal(pixel_list, required_pixels))
 
 
-def anndata_to_image_array(adata: AnnData) -> np.ndarray:
+def anndata_to_image_array(adata: AnnData, layer: Optional[str]=None) -> np.ndarray:
```
Comment (tdrose) [inline metaspace_converter/anndata_to_array.py:20]: Also added `layer` argument to the `anndata_to_image_array` function. With this also the processed images from layers can be extracted.

File: metaspace_converter/colocalization.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        layer: Key for the adata.layers dict in which the processed data will be saved.
+            Default value is ``colocml_preprocessing``.
+            If None, ``adata.X`` will be overwritten with the processed data.
+        median_filter_size: 2-dimensional filter size for the median filtering that
+            will be performed per ion image.
+        quantile_threshold: Float between 0 and 1. The function computes the quantile value per
+            ion image and all pixels below the quantile threshold will be set to 0.
+
+    Returns:
+        None. The processed data is saved in ``layer``. If layer is net to None, ``adata.X`` will
```
Comment (aeisenbarth) [inline metaspace_converter/colocalization.py:41]: ~net~ set

**Review by aeisenbarth — APPROVED** [review]
Besides one typo ready to merge!

**Pull request merged**